sshd-sftp/src/main/java/org/apache/sshd/client/subsystem/sftp/impl/SftpInputStreamAsync.java

```diff
@@ -0,0 +1,312 @@
+        }
+        int idx = off;
+        while (len > 0 && !eofIndicator) {
+            if (hasNoData()) {
+                fillData();
+                if (eofIndicator && (hasNoData())) {
+                    break;
+                }
+                sendRequests();
+            } else {
+                int nb = Math.min(buffer.available(), len);
+                buffer.getRawBytes(b, off, nb);
+                idx += nb;
+                len -= nb;
+                clientOffset += nb;
+            }
+        }
+        int res = idx - off;
+        if (res == 0 && eofIndicator) {
+            res = -1;
+        }
+        return res;
+    }
+
+    public long transferTo(long max, WritableByteChannel out) throws IOException {
+        if (!isOpen()) {
+            throw new IOException("transferTo(" + getPath() + ") stream closed");
+        }
+        long orgOffset = clientOffset;
+        while (!eofIndicator && max > 0) {
+            if (hasNoData()) {
+                fillData();
+                if (eofIndicator && hasNoData()) {
+                    break;
+                }
+                sendRequests();
+            } else {
+                int nb = buffer.available();
+                int toRead = (int) Math.min(nb, max);
+                ByteBuffer bb = ByteBuffer.wrap(buffer.array(), buffer.rpos(), toRead);
+                while (bb.hasRemaining()) {
+                    out.write(bb);
+                }
+                buffer.rpos(buffer.rpos() + toRead);
+                clientOffset += toRead;
+                max -= toRead;
+            }
+        }
+        return clientOffset - orgOffset;
+    }
+
+    @SuppressWarnings("PMD.MissingOverride")
+    public long transferTo(OutputStream out) throws IOException {
+        if (!isOpen()) {
+            throw new IOException("transferTo(" + getPath() + ") stream closed");
+        }
+        long orgOffset = clientOffset;
+        while (!eofIndicator) {
+            if (hasNoData()) {
+                fillData();
+                if (eofIndicator && hasNoData()) {
+                    break;
+                }
+                sendRequests();
+            } else {
+                int nb = buffer.available();
+                out.write(buffer.array(), buffer.rpos(), nb);
+                buffer.rpos(buffer.rpos() + nb);
+                clientOffset += nb;
+            }
+        }
+        return clientOffset - orgOffset;
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
+        if (!isOpen()) {
+            throw new IOException("skip(" + getPath() + ") stream closed");
+        }
+        if (clientOffset == 0 && pendingReads.isEmpty()) {
+            clientOffset = n;
+            return n;
+        }
+        return super.skip(n);
+    }
+
+    boolean hasNoData() {
+        return buffer == null || buffer.available() == 0;
+    }
+
+    void sendRequests() throws IOException {
```

Comment:
Let's make it protected in accordance with our policy to allow maximum flexibility to users who wish to modify/enhance the behavior of the relevant class